server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthAuthenticationManagerIT.java

```diff
@@ -842,11 +845,60 @@ void rejectTokenWithInvalidSignature() {
+        configureTokenKeyResponse("http://localhost/token_key", response, twice());
         assertThrows(InvalidTokenException.class, () -> externalOAuthAuthenticationManager.authenticate(xCodeToken));
     }
 
+    @Test
+    void updateKeyAtTheProviderWhileUAAHasOldKeyCached() throws Exception {
+        config.setTokenKeyUrl(new URL("http://localhost/token_key"));
+        config.setTokenKey(null);
+        String firstKey = invalidRsaSigningKey;
+        String secondKey = PRIVATE_KEY;
+
+        //first set of mocking to assure retrieval of first key from provider
+        KeyInfo key = KeyInfoBuilder.build("my-key-id", firstKey, UAA_ISSUER_URL);
+        VerificationKeyResponse verificationKeyResponse = TokenKeyEndpoint.getVerificationKeyResponse(key);
+        String response = JsonUtils.writeValueAsString(verificationKeyResponse);
+        signer = KeyInfoBuilder.build("my-key-id", firstKey, UAA_ISSUER_URL).getSigner();
+        mockToken();
+        mockUaaServer.expect(requestTo("http://localhost/token_key"))
+                .andExpect(header("Authorization", "Basic " + new String(Base64.encodeBase64("identity:identitysecret".getBytes()))))
+                .andExpect(header("Accept", "application/json"))
+                .andRespond(withStatus(OK).contentType(APPLICATION_JSON).body(response));
+
+        UaaUser existingShadowUser = new UaaUser(new UaaUserPrototype()
+                .withUsername("12345")
+                .withPassword("")
+                .withEmail("marissa_old@bloggs.com")
+                .withGivenName("Marissa_Old")
+                .withFamilyName("Bloggs_Old")
+                .withId("user-id")
+                .withOrigin("the_origin")
+                .withZoneId("uaa")
+                .withAuthorities(UaaAuthority.USER_AUTHORITIES));
+
+        userDatabase.addUser(existingShadowUser);
+
+        //another set of mocking to assure retrieval of second key from provider
+        //System should trigger retrieving of seconding key after invalidating the cache.
+        key = KeyInfoBuilder.build("my-key-id", secondKey, UAA_ISSUER_URL);
+        verificationKeyResponse = TokenKeyEndpoint.getVerificationKeyResponse(key);
+        response = JsonUtils.writeValueAsString(verificationKeyResponse);
+        signer = KeyInfoBuilder.build("my-key-id", secondKey, UAA_ISSUER_URL).getSigner();
+        mockToken();
+        mockUaaServer.expect(requestTo("http://localhost/token_key"))
+                .andExpect(header("Authorization", "Basic " + new String(Base64.encodeBase64("identity:identitysecret".getBytes()))))
+                .andExpect(header("Accept", "application/json"))
+                .andRespond(withStatus(OK).contentType(APPLICATION_JSON).body(response));
+
+        externalOAuthAuthenticationManager.authenticate(xCodeToken); //with first key
+        externalOAuthAuthenticationManager.authenticate(xCodeToken); //with second key
```

Comment:
And then here it's gonna used the previously cached key (aka the second key), and it works.